File: Player/PlayerControl.js
// This script control the player's movements.
// These variables control the player's moving and jumping speed.
var runSpeed = 7;
var jumpSpeed = 10;

// This variable is used to control if the player is allowed to move or not.
var playerCanMove = true;

// This variable is used to check if the player is in contact with the ground.
var isGrounded : boolean;

// This variable is used by the attack script and for animation.
var isAttacking : boolean;

// This variable is used to control the orientation of the trigger collider.
var triggerCol : Collider2D;

// This variable is used in the NPCs Dialog management part.
var NPC : GameObject;

// This part define the character's controls.
function Update () {
    PlayerMove();
}

function PlayerMove(){
    var rb2d = GetComponent(Rigidbody2D); 
    var sprite = GetComponent(SpriteRenderer);
    var move = Vector2(Input.GetAxis("Horizontal"), 0);

    if (playerCanMove) {
        transform.position += move * runSpeed * Time.deltaTime;

        if (Input.GetButtonDown("Jump") && isGrounded) {
            rb2d.velocity = new Vector2(0.0f, jumpSpeed);
        }

        if (Input.GetButtonDown("Attack")) {
            Debug.Log("Player attacked.");
            isAttacking = true;
        }

    // This part flip the player's sprite and the two BoxCollider2D in the direction he moves.
        if ((Input.GetAxis("Horizontal") > 0)) {
            sprite.flipX = false;
            GetComponent(BoxCollider2D).offset.x = 0.06;
            triggerCol.offset.x = 0.06;  
        }

        if ((Input.GetAxis("Horizontal") < 0)) {
            sprite.flipX = true;
            GetComponent(BoxCollider2D).offset.x = -0.06;
            triggerCol.offset.x = -0.06;
        }
    }

    if (sprite.flipX) {
        GameObject.Find("Ghost").GetComponent(SpriteRenderer).flipX = true;        
    }
    else {
        GameObject.Find("Ghost").GetComponent(SpriteRenderer).flipX = false;        
    }
}

function OnTriggerStay2D (other : Collider2D) {
// This part acquire the NPC who's in contact with the player.
    if (other.gameObject.tag == "NPC") {
        NPC = other.gameObject;
    }

// This part check if the player is on the ground.
    if (other.gameObject.tag == "Ground") {
        isGrounded = true;       
    }
}

function OnTriggerExit2D (other: Collider2D) {
// This part check if the player left the ground.    
    if (other.gameObject.tag == "Ground") {
        isGrounded = false;
    }
}

File: Player/PlayerControl.cs
using UnityEngine;

public class PlayerControl : MonoBehaviour {
	[SerializeField] float runSpeed = 18;
	[SerializeField] float jumpSpeed = 18;

	public GameObject npc;
	public bool canMove = false;
	public bool isGrounded = false;

	GameObject ghost;
	GameObject playerUI;
	Animator animator;
	AudioSource audioSource;
	Rigidbody2D rb;
	SpriteRenderer sprite;
	PlayerInputActions playerInputs;
	SceneTransition SceneTransition;
	PauseMenuManager pauseMenuManager;
	PlayerHealth playerHealth;
	PlayerSound playerSound;

	bool jump = false;
	float direction = 0;
	float colliderOffsetX = 0;
	float colliderOffsetY = 0;

	void Awake() {
		playerInputs = new PlayerInputActions();
		playerInputs.Player.Move.performed += ctx => direction = ctx.ReadValue<float>();
		playerInputs.Player.Jump.performed += ctx => jump = true;
		playerInputs.Player.Jump.canceled += ctx => jump = false;
	}

	void Start() {
		playerUI = GameObject.Find("PlayerUI");
		ghost = GameObject.Find("Ghost");
		rb = GetComponent<Rigidbody2D>();
		sprite = GetComponent<SpriteRenderer>();
		animator = GetComponent<Animator>();
		audioSource = GetComponent<AudioSource>();
		playerHealth = GetComponent<PlayerHealth>();
		playerSound = GetComponent<PlayerSound>();
		colliderOffsetX = GetComponent<CapsuleCollider2D>().offset.x;
		colliderOffsetY = GetComponent<CapsuleCollider2D>().offset.y;
		SceneTransition = GameObject.Find("SceneTransition").GetComponent<SceneTransition>();
		pauseMenuManager = GameObject.Find("PlayerUI").GetComponent<PauseMenuManager>();
	}

	void Update() {
		bool isLoading = SceneTransition ? SceneTransition.isLoading : false;
		bool inDialog = playerUI ? playerUI.GetComponent<DialogManager>().inDialog : false;
		bool tookDamage = playerHealth ? playerHealth.isInv : false;
		bool isDead = playerHealth ? playerHealth.isDead : false;
		bool isPaused = pauseMenuManager? pauseMenuManager.paused : false;
		bool introDone = PlayerState.Instance.introDone;

		canMove = !isPaused && !inDialog && !isLoading && !tookDamage && !isDead && introDone;

		if (inDialog || isLoading) GetComponent<Rigidbody2D>().velocity = new Vector2(0, 0);
		animator.SetBool("run", GetComponent<Rigidbody2D>().velocity.x != 0);
		animator.SetBool("air", !isGrounded);

		GetComponent<CapsuleCollider2D>().offset = new Vector2(sprite.flipX ? -colliderOffsetX : colliderOffsetX, colliderOffsetY);
		if (ghost) ghost.GetComponent<SpriteRenderer>().flipX = sprite.flipX;
	}

	void FixedUpdate() {
		if (canMove) {
			rb.velocity = new Vector2(direction * runSpeed, rb.velocity.y);
			if (rb.velocity.x > 0) transform.eulerAngles = new Vector3(0, 0, 0);
			if (rb.velocity.x < 0) transform.eulerAngles = new Vector3(0, 180, 0);
			if (jump && isGrounded && playerSound && npc == null) {
				rb.velocity = (Vector2.up * jumpSpeed);
				audioSource.PlayOneShot(playerSound.jumpSound);
				jump = false;
			}
		}
	}

	void OnCollisionStay2D(Collision2D other) {
		if (other.gameObject.tag == "Ground") isGrounded = true;
	}

	void OnCollisionExit2D(Collision2D other) {
		if (other.gameObject.tag == "Ground") isGrounded = false;
	}

	void OnTriggerEnter2D(Collider2D other) {
		if (other.gameObject.tag == "NPC" || other.gameObject.tag == "StaticNPC") npc = other.gameObject;
	}

	void OnTriggerExit2D(Collider2D other) {
		npc = null;
		if (other.tag == "Level") StartCoroutine(SceneTransition.LoadScene(false));
	}

	void OnEnable() {
		playerInputs.Enable();
	}

	void OnDisable() {
		playerInputs.Disable();
	}
}